node/impl/full/state.go

```diff
@@ -1436,72 +1436,93 @@ func (m *StateModule) MsigGetPending(ctx context.Context, addr address.Address,
 	} else {
-		sectorWeight = miner.QAPowerMax(ssize)
+		// NB: not exactly accurate, but should always lead us to *over* estimate, not under
+		duration := pci.Expiration - height
+		sectorWeight = builtin.QAPowerForWeight(ssize, duration, w, vw)
 	}
 
-	var powerSmoothed builtin.FilterEstimate
+	return sectorWeight, nil
+}
+
+func (a *StateAPI) pledgeCalculationInputs(ctx context.Context, state *state.StateTree) (abi.TokenAmount, *builtin.FilterEstimate, error) {
+	store := a.Chain.ActorStore(ctx)
+
+	var (
+		powerSmoothed    builtin.FilterEstimate
+		pledgeCollateral abi.TokenAmount
+	)
 	if act, err := state.GetActor(power.Address); err != nil {
-		return types.EmptyInt, xerrors.Errorf("loading power actor: %w", err)
+		return types.EmptyInt, nil, xerrors.Errorf("loading power actor: %w", err)
 	} else if s, err := power.Load(store, act); err != nil {
-		return types.EmptyInt, xerrors.Errorf("loading power actor state: %w", err)
+		return types.EmptyInt, nil, xerrors.Errorf("loading power actor state: %w", err)
 	} else if p, err := s.TotalPowerSmoothed(); err != nil {
-		return types.EmptyInt, xerrors.Errorf("failed to determine total power: %w", err)
+		return types.EmptyInt, nil, xerrors.Errorf("failed to determine total power: %w", err)
+	} else if c, err := s.TotalLocked(); err != nil {
+		return types.EmptyInt, nil, xerrors.Errorf("failed to determine pledge collateral: %w", err)
 	} else {
 		powerSmoothed = p
+		pledgeCollateral = c
+	}
+
+	return pledgeCollateral, &powerSmoothed, nil
+}
+
+func (a *StateAPI) StateMinerPreCommitDepositForPower(ctx context.Context, maddr address.Address, pci miner.SectorPreCommitInfo, tsk types.TipSetKey) (types.BigInt, error) {
+	ts, err := a.Chain.GetTipSetFromKey(ctx, tsk)
+	if err != nil {
+		return types.EmptyInt, xerrors.Errorf("loading tipset %s: %w", tsk, err)
+	}
+
+	state, err := a.StateManager.ParentState(ts)
+	if err != nil {
+		return types.EmptyInt, xerrors.Errorf("loading state %s: %w", tsk, err)
 	}
 
 	rewardActor, err := state.GetActor(reward.Address)
 	if err != nil {
-		return types.EmptyInt, xerrors.Errorf("loading miner actor: %w", err)
+		return types.EmptyInt, xerrors.Errorf("loading reward actor: %w", err)
 	}
 
-	rewardState, err := reward.Load(store, rewardActor)
+	rewardState, err := reward.Load(a.Chain.ActorStore(ctx), rewardActor)
 	if err != nil {
 		return types.EmptyInt, xerrors.Errorf("loading reward actor state: %w", err)
 	}
 
-	deposit, err := rewardState.PreCommitDepositForPower(powerSmoothed, sectorWeight)
+	sectorWeight, err := a.calculateSectorWeight(ctx, maddr, pci, ts.Height(), state)
 	if err != nil {
-		return big.Zero(), xerrors.Errorf("calculating precommit deposit: %w", err)
+		return types.EmptyInt, err
+	}
+
+	_, powerSmoothed, err := a.pledgeCalculationInputs(ctx, state)
+	if err != nil {
+		return types.EmptyInt, err
+	}
+
+	deposit, err := rewardState.PreCommitDepositForPower(*powerSmoothed, sectorWeight)
```

Comment:
sorry @ZenGround0 I'm struggling to follow this comment or figure out how to act on it, can you explain further and demonstrate how it might lead to a refactor of what's in here already? I'm not really up to speed on reward smoothed and power smoothed, so this is just monkey-typey refactoring without changing that logic.

Reply:
in that case dont worry about it